src/plugin/ControlBar/AnimationProgressBar.ts

```diff
@@ -0,0 +1,334 @@
+    window.removeEventListener(BROWSER.EVENTS.MOUSE_MOVE, this._onMouseMove, false);
+    window.removeEventListener(BROWSER.EVENTS.MOUSE_UP, this._onMouseUp, false);
+    this._hideKnob();
+    this._setAnimationTimeScale(this._origTimeScale);
+  };
+
+  private _onTouchStart = (evt: TouchEvent) => {
+    if (evt.touches.length > 1) return;
+
+    const touch = evt.touches[0];
+    const animator = this._view3D.animator;
+    const activeAnimationIdx = animator.activeAnimationIndex;
+    const activeAnimationAction = animator.actions[activeAnimationIdx];
+
+    this._showKnob();
+    this._firstTouch = { x: touch.pageX, y: touch.pageY };
+    this._origTimeScale = activeAnimationAction.getEffectiveTimeScale();
+    this._setAnimationTimeScale(0);
+    this._updateAnimationProgress(touch.pageX);
+  };
+
+  private _onTouchMove = (evt: TouchEvent) => {
+    // Only the one finger motion should be considered
+    if (evt.touches.length > 1 || this._scrolling) return;
+
+    const touch = evt.touches[0];
+    const scrollable = this._view3D.scrollable;
+    const firstTouch = this._firstTouch;
+
+    if (firstTouch) {
+      if (scrollable) {
+        const delta = new THREE.Vector2(touch.pageX, touch.pageY)
+          .sub(new THREE.Vector2(firstTouch.x, firstTouch.y));
+
+        if (Math.abs(delta.y) > Math.abs(delta.x)) {
+          // Assume Scrolling
+          this._scrolling = true;
+          this._release();
+          return;
+        }
+      }
+
+      this._firstTouch = null;
+    }
+
+    if (evt.cancelable) {
+      evt.preventDefault();
+    }
+
+    evt.stopPropagation();
+    this._setAnimationTimeScale(0);
+    this._updateAnimationProgress(touch.pageX);
+  };
+
+  private _onTouchEnd = (evt: TouchEvent) => {
+    if (evt.touches.length > 0) return;
+
+    this._release();
+    this._scrolling = false;
+  };
+
+  private _release() {
+    this._hideKnob();
+    this._setAnimationTimeScale(this._origTimeScale);
+  }
+
+  private _showKnob() {
+    const knob = this._knobEl;
+    const controlBar = this._controlBar;
+    const className = {
+      ...controlBar.className,
+      ...ControlBar.DEFAULT_CLASS
+    };
+
+    knob.classList.add(className.VISIBLE);
+  }
+
+  private _hideKnob() {
+    const knob = this._knobEl;
+    const controlBar = this._controlBar;
+    const className = {
+      ...controlBar.className,
+      ...ControlBar.DEFAULT_CLASS
+    };
+
+    knob.classList.remove(className.VISIBLE);
+  }
+
+  private _updateAnimationProgress = (x: number) => {
+    const view3D = this._view3D;
+    const rootBbox = this._rootBbox;
+    const knob = this._knobEl;
+
+    const animator = view3D.animator;
+    const activeAnimationIdx = animator.activeAnimationIndex;
+    const activeAnimationClip = animator.activeAnimation;
+    const activeAnimationAction = animator.actions[activeAnimationIdx];
+
+    if (!activeAnimationClip || !activeAnimationAction) return;
+
+    const progress = ((x - rootBbox.x) / rootBbox.width);
```

Comment:
If there is a scroll left, would the value be different?

Reply:
Good point, I checked & fixed the wrong behavior of the animation progress bar when there's a scroll bar that exists outside the canvas area.